src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,21 +124,75 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
     }
 
     qint64 offset = 0;
+    QByteArray protrudingData;
+    bool protrudingDataLeft = false;
 
     QObject::connect(&socket, &QLocalSocket::readyRead, &loop, [&] {
-        auto data = socket.readAll();
+        const auto data = socket.readAll();
         if (data.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        QByteArray dataToSent;
+        // CFAPI expectes sended blocks to be of a specified size.
+        // Only the last sended block is allowed to be smaller than the specified block size
+        // and therefore marks the end of the transmission.
+        static constexpr auto CFAPI_BLOCK_SIZE = 4096;
+
+        int dataIndex = 0;
+        // Is there any data left from last run?
+        if (!protrudingData.isEmpty()) {
+            QByteArray dataToSent;
+            dataToSent.append(protrudingData);
+
+            // Fill up remaining first block if possible
+            const auto fillupSize = CFAPI_BLOCK_SIZE - dataToSent.size();
+            if (fillupSize > 0) {
+                dataToSent.append(data.left(fillupSize));
+                dataIndex += fillupSize;
+            }
+
+            // Send the block
+            qDebug(lcCfApiWrapper) << "Send protruding data that is left from last run. Size:" << dataToSent.size();
+            sendTransferInfo(dataToSent, offset);
+            offset += dataToSent.length();
+        }
+
+        // Do we still have data to sent?
+        if (dataIndex < data.size()) {
+            const auto blocksCount = (data.size() - (dataIndex + 1)) / CFAPI_BLOCK_SIZE;
+            if (blocksCount != 0) {
+                // Send the blocks one by one
+                for (int i = 0; i < blocksCount; ++i) {
+                    auto dataToSent = data.mid(dataIndex, CFAPI_BLOCK_SIZE);
+                    qDebug(lcCfApiWrapper) << "Send data block. Size:" << dataToSent.size();
+                    sendTransferInfo(dataToSent, offset);
```

Comment:
@FlexW I've been also thinking (well, because, I found the logic a bit complicated to comprehend for me at one go, or maybe it's because it's too late in the evening:) ), wouldn't it be easier to just accumulate the incoming data from the **data** variable to a **dataToSend** variable, no matter how many data comes in. Just accumulate it in such a way that it will always get stored in one piece as much as multiples of **4096** it can allow. Then, the rest of the data will go into that **protrudingData** variable.
So, let's say, we have **data.size()** of 17408KB. Can we just extract blocks that are multiples of 4096 into the **dataToSent*, until we've left with just 1024KB of data that is not multiple of 4096. Then, we can have the `sendTransferInfo(dataToSend, offset)` call only in one place, and then, we can keep the remaining 1024KB for the next iteration?